File: ascendCBL/PLOTS.R
# ======================================================================
# make_figure.R
#
# Build the headline ASCEND vs CBL figure from results.csv.
#
# Usage:
#   Rscript make_figure.R                       # reads ./results.csv
#   Rscript make_figure.R path/to/results.csv   # explicit path
#
# Outputs (next to the CSV by default):
#   ascend_vs_cbl.pdf
#   ascend_vs_cbl.png
#
# Packages: ggplot2, patchwork, data.table, scales, grid
# ======================================================================

suppressPackageStartupMessages({
  library(data.table)
  library(ggplot2)
  library(patchwork)
  library(scales)
  library(grid)
})

args     <- commandArgs(trailingOnly = TRUE)
CSV_PATH <- if (length(args) >= 1) args[1] else "results.csv"
OUT_DIR  <- if (length(args) >= 2) args[2] else dirname(normalizePath(CSV_PATH))
stopifnot(file.exists(CSV_PATH))

# ---- Palette / theme ---------------------------------------------------
C_ASCEND  <- "#0072B2"   # blue
C_CBL     <- "#E69F00"   # orange
C_TIMEOUT <- "#C44E52"   # red
GREY      <- "#5A5A5A"

method_levels <- c("CBL (baseline)", "ASCEND (ours)")
method_pal    <- c("CBL (baseline)" = C_CBL, "ASCEND (ours)" = C_ASCEND)

theme_paper <- function(base = 10) {
  theme_bw(base_size = base) +
    theme(
      panel.grid.minor   = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.major.y = element_line(linewidth = 0.25, colour = "grey85"),
      panel.border       = element_blank(),
      axis.line          = element_line(linewidth = 0.45, colour = "grey20"),
      axis.ticks         = element_line(linewidth = 0.45, colour = "grey20"),
      plot.title         = element_text(face = "bold", size = base + 1.5,
                                        margin = margin(b = 4)),
      plot.title.position = "plot",
      strip.background   = element_blank(),
      legend.position    = "none"   # one shared legend on the figure
    )
}

# ---- Load & label ------------------------------------------------------
d <- fread(CSV_PATH)
d[, timed_out := status == "timeout"]
d[, method_lbl := factor(
  ifelse(method == "ascend", "ASCEND (ours)", "CBL (baseline)"),
  levels = method_levels)]

ok <- d[status == "ok"]

# Aggregator: mean and SEM over seeds.
agg <- function(df, group_cols, metric) {
  setDT(df)
  df[, .(mean = mean(get(metric), na.rm = TRUE),
         sd   = sd(get(metric),   na.rm = TRUE),
         n_   = sum(!is.na(get(metric)))),
     by = group_cols][
       , sem := sd / sqrt(pmax(n_, 1))][]
}

# Slice helpers — three 1-D sweeps through the default point.
slice_dx <- d[n == 1024 & d_z == 10]
slice_dz <- d[n == 1024 & d_x == 5]

# ---- Generic sweep panel ----------------------------------------------
# `metric` along `axis`. ASCEND/CBL ribbons = ±1 SEM. If `show_timeouts`
# is TRUE we overlay red ✕ at the mean censored time for cells where
# CBL timed out.
sweep_panel <- function(df, axis, metric, ylab, title,
                        log_y = FALSE, show_timeouts = FALSE,
                        x_breaks = NULL, x_label = NULL,
                        annot = NULL) {
  
  okd <- df[status == "ok"]
  a   <- agg(okd, c("method_lbl", axis), metric)
  setnames(a, axis, "x")
  a <- a[!is.na(mean)]
  
  p <- ggplot(a, aes(x = x, y = mean, colour = method_lbl, fill = method_lbl,
                     group = method_lbl)) +
    geom_ribbon(aes(ymin = pmax(mean - sem, if (log_y) 1e-9 else -Inf),
                    ymax = mean + sem),
                alpha = 0.18, colour = NA) +
    geom_line(linewidth = 1.0) +
    geom_point(size = 2.4, shape = 21, colour = "white", stroke = 0.7) +
    scale_colour_manual(values = method_pal) +
    scale_fill_manual  (values = method_pal) +
    labs(title = title, x = x_label, y = ylab) +
    theme_paper()
  
  if (!is.null(x_breaks)) p <- p + scale_x_continuous(breaks = x_breaks)
  if (log_y)              p <- p + scale_y_log10(labels = label_comma())
  
  if (show_timeouts) {
    t <- df[timed_out == TRUE & method == "cbl"]
    if (nrow(t) > 0) {
      t_mean <- t[, .(y = mean(time_sec)), by = axis]
      setnames(t_mean, axis, "x")
      p <- p + geom_point(data = t_mean,
                          aes(x = x, y = y),
                          colour = C_TIMEOUT, shape = 4, size = 4,
                          stroke = 1.4, inherit.aes = FALSE)
    }
  }
  
  if (!is.null(annot)) {
    for (a_ in annot) {
      p <- p + annotate("text",
                        x = a_$x, y = a_$y, label = a_$label,
                        hjust = a_$hjust %||% 0, vjust = a_$vjust %||% 0.5,
                        colour = a_$colour, fontface = "bold", size = 3.4)
    }
  }
  p
}

`%||%` <- function(a, b) if (is.null(a)) b else a

# ---- (a) Runtime vs d_x ----------------------------------------------
p_a <- sweep_panel(
  slice_dx, "d_x", "time_sec",
  ylab = "Wall-clock time (s)",
  title = expression(bold("(a) Runtime vs " * d[x])),
  log_y = TRUE, show_timeouts = TRUE,
  x_breaks = c(5, 10, 15, 20),
  x_label = expression(d[x] * "  (foreground variables)")
) +
  annotate("text", x = 5, y = Inf, label = "n=1024,  d_z=10",
           hjust = -0.05, vjust = 1.6, colour = GREY, size = 3.0)

# ---- (b) Runtime vs d_z ----------------------------------------------
# Compute the slope annotations from the data themselves.
slope_label <- function(method_name, axis_var, slice) {
  s <- slice[status == "ok" & method == method_name,
             .(t = mean(time_sec)), by = axis_var]
  setnames(s, axis_var, "x")
  s <- s[!is.na(t) & t > 0]
  if (nrow(s) < 2) return(NA_real_)
  unname(coef(lm(log(t) ~ log(x), data = s))[2])
}
slope_cbl_dz    <- slope_label("cbl",    "d_z", slice_dz)
slope_ascend_dz <- slope_label("ascend", "d_z", slice_dz)

p_b <- sweep_panel(
  slice_dz, "d_z", "time_sec",
  ylab = "Wall-clock time (s)",
  title = expression(bold("(b) Runtime vs " * d[z])),
  log_y = TRUE, show_timeouts = TRUE,
  x_breaks = c(10, 20, 30, 40, 50),
  x_label = expression(d[z] * "  (background variables)")
) +
  annotate("text", x = 10, y = Inf, label = "n=1024,  d_x=5",
           hjust = -0.05, vjust = 1.6, colour = GREY, size = 3.0) +
  annotate("text", x = 35, y = 200,
           label = sprintf("CBL: t \u221d d_z^{%+0.2f}", slope_cbl_dz),
           colour = C_CBL, fontface = "bold", hjust = 0, size = 3.4) +
  annotate("text", x = 20, y = 0.08,
           label = sprintf("ASCEND: t \u221d d_z^{%+0.2f}", slope_ascend_dz),
           colour = C_ASCEND, fontface = "bold", hjust = 0, size = 3.4)

# ---- (c) Speedup factor bar chart ------------------------------------
# Use mean time per (method, axis-value) including timeouts.
speedup_along <- function(slice, axis) {
  a <- slice[, .(t = mean(time_sec, na.rm = TRUE)), by = c("method", axis)]
  w <- dcast(a, as.formula(paste(axis, "~ method")), value.var = "t")
  w[, speedup := cbl / ascend]
  w
}
sp_dx <- speedup_along(slice_dx, "d_x")
sp_dz <- speedup_along(slice_dz, "d_z")

# Which axis values contained a CBL timeout?
to_dx <- unique(slice_dx[timed_out == TRUE]$d_x)
to_dz <- unique(slice_dz[timed_out == TRUE]$d_z)

bars <- rbind(
  sp_dx[, .(label = paste0("d_x=", d_x),
            value = speedup,
            group = "varying d_x",
            timed_out = d_x %in% to_dx,
            order_key = d_x)],
  sp_dz[, .(label = paste0("d_z=", d_z),
            value = speedup,
            group = "varying d_z",
            timed_out = d_z %in% to_dz,
            order_key = 100 + d_z)]   # offset so d_z bars come after d_x bars
)
bars[, label := factor(label, levels = label[order(order_key)])]

bar_pal <- c("varying d_x" = "#7FB3D5", "varying d_z" = "#1f6f9c")

p_c <- ggplot(bars, aes(x = label, y = value, fill = group)) +
  geom_col(width = 0.78,
           colour = ifelse(bars$timed_out, C_TIMEOUT, "white"),
           linewidth = ifelse(bars$timed_out, 1.1, 0.5)) +
  geom_text(aes(label = paste0(format(round(value), big.mark = ","), "\u00d7")),
            vjust = -0.4, size = 3.3, colour = "grey20") +
  geom_hline(yintercept = 1, linetype = "dashed", colour = GREY,
             linewidth = 0.4) +
  scale_y_log10(labels = label_comma(),
                expand = expansion(mult = c(0.05, 0.22))) +
  scale_fill_manual(values = bar_pal) +
  labs(title = "(c) ASCEND speedup factor",
       x = NULL,
       y = "Speedup  (CBL time / ASCEND time)") +
  theme_paper() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8.5),
        legend.position = c(0.02, 0.98),
        legend.justification = c(0, 1),
        legend.title = element_blank(),
        legend.background = element_rect(fill = alpha("white", 0.7),
                                         colour = NA),
        legend.key.size = unit(0.4, "cm"),
        legend.text  = element_text(size = 8.5))

# Mark the x-axis tick labels of timeout bars in red, and add a small
# explanatory note inside the panel.
if (any(bars$timed_out)) {
  tick_cols <- ifelse(levels(bars$label) %in%
                        as.character(bars$label[bars$timed_out]),
                      C_TIMEOUT, "grey20")
  p_c <- p_c +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8.5,
                                     colour = tick_cols)) +
    annotate("text", x = 1, y = max(bars$value) * 2.4,
             label = "red border / red tick = CBL hit timeout\n(speedup is a lower bound)",
             hjust = 0, vjust = 1, size = 2.8, colour = C_TIMEOUT,
             fontface = "italic")
}

# ---- (d) Computational work (CI tests) -------------------------------
p_d <- sweep_panel(
  slice_dz, "d_z", "n_ci_tests",
  ylab = "Conditional independence\ntests performed",
  title = "(d) Computational work",
  log_y = TRUE,
  x_breaks = c(10, 20, 30, 40, 50),
  x_label = expression(d[z])
) +
  annotate("text", x = 10, y = Inf, label = "n=1024,  d_x=5",
           hjust = -0.05, vjust = 1.6, colour = GREY, size = 3.0)

# ---- (e) Decisiveness ------------------------------------------------
p_e <- sweep_panel(
  slice_dz, "d_z", "coverage",
  ylab = "Coverage  (1 \u2212 NA fraction)",
  title = "(e) Decisiveness",
  x_breaks = c(10, 20, 30, 40, 50),
  x_label = expression(d[z])
) +
  coord_cartesian(ylim = c(0.45, 1.03)) +
  annotate("text", x = 10, y = 0.47, label = "n=1024,  d_x=5",
           hjust = -0.05, vjust = 0, colour = GREY, size = 3.0)

# ---- (f) Accuracy vs d_x --------------------------------------------
p_f <- sweep_panel(
  slice_dx, "d_x", "f1",
  ylab = "F1 score",
  title = expression(bold("(f) Accuracy vs " * d[x])),
  x_breaks = c(5, 10, 15, 20),
  x_label = expression(d[x])
) +
  coord_cartesian(ylim = c(-0.02, 1.03)) +
  annotate("text", x = 5, y = -0.02, label = "n=1024,  d_z=10",
           hjust = -0.05, vjust = -0.4, colour = GREY, size = 3.0) +
  # Call out the missing CBL point at d_x=5
  annotate("text", x = 5.3, y = 0.05,
           label = "CBL: 0 TPs across\nall 5 seeds (F1 undef.)",
           hjust = 0, vjust = 0, size = 2.8, colour = C_CBL,
           fontface = "italic")

# ---- Assemble & save -------------------------------------------------
# Custom shared legend strip across the top.
top_legend <- ggplot() +
  annotate("segment", x = 0.06, xend = 0.10, y = 0.5, yend = 0.5,
           colour = C_ASCEND, linewidth = 1.2) +
  annotate("point",   x = 0.08,             y = 0.5,
           colour = C_ASCEND, fill = C_ASCEND, shape = 21, size = 3.0,
           stroke = 0.9) +
  annotate("text",    x = 0.115, y = 0.5, hjust = 0,
           label = "ASCEND (ours)", size = 3.6) +
  annotate("segment", x = 0.30, xend = 0.34, y = 0.5, yend = 0.5,
           colour = C_CBL, linewidth = 1.2) +
  annotate("point",   x = 0.32,             y = 0.5,
           colour = C_CBL, fill = C_CBL, shape = 21, size = 3.0, stroke = 0.9) +
  annotate("text",    x = 0.355, y = 0.5, hjust = 0,
           label = "CBL (baseline)", size = 3.6) +
  annotate("point",   x = 0.555, y = 0.5,
           colour = C_TIMEOUT, shape = 4, size = 4, stroke = 1.5) +
  annotate("text",    x = 0.575, y = 0.5, hjust = 0,
           label = "CBL exceeded 1 h timeout", size = 3.6) +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, 1), expand = FALSE) +
  theme_void()

fig <- (p_a | p_b | p_c) / (p_d | p_e | p_f) &
  theme(plot.margin = margin(6, 8, 6, 8))

fig <- top_legend / fig +
  plot_layout(heights = c(0.06, 1)) +
  plot_annotation(
    title = "ASCEND vs CBL  \u2014  scaling, computational cost, and discovery quality  (5 seeds, ribbons = \u00b11 SEM)",
    theme = theme(plot.title = element_text(face = "bold", size = 12.5,
                                            hjust = 0.5,
                                            margin = margin(b = 4)))
  )

pdf_path <- file.path(OUT_DIR, "ascend_vs_cblRR.pdf")
png_path <- file.path(OUT_DIR, "ascend_vs_cblRR.png")
ggsave(pdf_path, fig, width = 13, height = 7.4, device = cairo_pdf)
ggsave(png_path, fig, width = 13, height = 7.4, dpi = 300)

cat("Wrote:\n  ", pdf_path, "\n  ", png_path, "\n", sep = "")

File: ascendCBL/plots.py
#!/usr/bin/env python3
"""
Build the headline ASCEND vs CBL figure from results CSV.

Usage:
    python make_figure.py results.csv outdir/

Output:
    outdir/ascend_vs_cbl.pdf
    outdir/ascend_vs_cbl.png
"""
import sys
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
import matplotlib.lines as mlines
from matplotlib.patches import Patch

# ---- style ----------------------------------------------------------------
mpl.rcParams.update({
    "font.family": "DejaVu Sans",
    "font.size": 10,
    "axes.titlesize": 11.5,
    "axes.titleweight": "bold",
    "axes.labelsize": 10.5,
    "axes.spines.top": False,
    "axes.spines.right": False,
    "axes.linewidth": 0.9,
    "xtick.major.width": 0.9,
    "ytick.major.width": 0.9,
    "xtick.labelsize": 9.5,
    "ytick.labelsize": 9.5,
    "legend.fontsize": 10,
    "legend.frameon": False,
    "figure.dpi": 140,
    "savefig.dpi": 300,
    "savefig.bbox": "tight",
})

C_ASCEND = "#0072B2"   # blue
C_CBL    = "#E69F00"   # orange
C_TIMEOUT= "#C44E52"   # red
GREY     = "#5A5A5A"

CSV_PATH = sys.argv[1] if len(sys.argv) > 1 else "/mnt/user-data/uploads/results_12.csv"
OUT_DIR  = sys.argv[2] if len(sys.argv) > 2 else "/mnt/user-data/outputs"

d = pd.read_csv(CSV_PATH)
d["timed_out"] = (d.status == "timeout")

# For metric panels, only ok rows. For runtime/speedup, include timeouts
# (they ARE part of CBL's time cost in any honest comparison).
ok = d[d.status == "ok"].copy()

def agg(df, group_cols, metric):
    """mean and SEM over seeds."""
    g = df.groupby(group_cols)[metric].agg(["mean", "std", "count"]).reset_index()
    g["sem"] = g["std"] / np.sqrt(g["count"])
    return g


def sweep_panel(ax, df, axis, metric, log_y=False, ribbon=True,
                show_timeouts=False, timeout_y_for_mean=False):
    """Plot mean ± SEM ribbon for each method along `axis`."""
    for method, colour in [("cbl", C_CBL), ("ascend", C_ASCEND)]:
        s = df[df.method == method]
        if len(s) == 0:
            continue
        a = agg(s, axis, metric).dropna(subset=["mean"]).sort_values(axis)
        if len(a) == 0:
            continue
        if ribbon and "sem" in a.columns:
            lo = a["mean"] - a["sem"]
            hi = a["mean"] + a["sem"]
            if log_y:
                lo = np.maximum(lo, 1e-9)
            ax.fill_between(a[axis], lo, hi, color=colour, alpha=0.18, lw=0)
        ax.plot(a[axis], a["mean"], "-", color=colour, lw=2.2, zorder=3)
        ax.plot(a[axis], a["mean"], "o", color=colour, ms=6.0,
                markeredgecolor="white", markeredgewidth=1.0, zorder=4)
    # Mark timeouts (CBL only in practice) with a red ✕ at the timeout time
    if show_timeouts:
        t = df[(df.timed_out) & (df.method == "cbl")]
        for xv, sub in t.groupby(axis):
            y = sub.time_sec.mean()
            ax.scatter([xv], [y], marker="x", color=C_TIMEOUT, s=85, lw=2.4,
                       zorder=6)
    if log_y:
        ax.set_yscale("log")


# Slices through default point
slice_dx = d[(d.n == 1024) & (d.d_z == 10)]
slice_dz = d[(d.n == 1024) & (d.d_x == 5)]


# ---- figure --------------------------------------------------------------
fig = plt.figure(figsize=(12.5, 7.4))
gs = fig.add_gridspec(2, 3, hspace=0.50, wspace=0.32,
                      left=0.06, right=0.985, top=0.88, bottom=0.08)
axes = [[fig.add_subplot(gs[r, c]) for c in range(3)] for r in range(2)]

# =========================== ROW 1 : SPEED ================================
# (a) Runtime vs d_x
ax = axes[0][0]
sweep_panel(ax, slice_dx, "d_x", "time_sec", log_y=True, show_timeouts=True)
ax.set_xlabel(r"$d_x$  (foreground variables)")
ax.set_ylabel("Wall-clock time (s)")
ax.set_title(r"(a) Runtime vs $d_x$")
ax.set_xticks([5, 10, 15, 20])
ax.grid(axis="y", which="major", alpha=0.25, lw=0.5)
ax.text(0.02, 0.97, r"n=1024,  $d_z$=10",
        transform=ax.transAxes, va="top", ha="left",
        fontsize=8.8, color=GREY)

# (b) Runtime vs d_z — the most dramatic scaling story
ax = axes[0][1]
sweep_panel(ax, slice_dz, "d_z", "time_sec", log_y=True, show_timeouts=True)
ax.set_xlabel(r"$d_z$  (background variables)")
ax.set_ylabel("Wall-clock time (s)")
ax.set_title(r"(b) Runtime vs $d_z$")
ax.set_xticks([10, 20, 30, 40, 50])
ax.grid(axis="y", which="major", alpha=0.25, lw=0.5)
ax.text(0.02, 0.97, r"n=1024,  $d_x$=5",
        transform=ax.transAxes, va="top", ha="left",
        fontsize=8.8, color=GREY)
# Scaling exponent annotations, positioned out of the way.
ax.text(0.97, 0.62, r"CBL: $t \propto d_z^{+0.81}$",
        transform=ax.transAxes, ha="right", va="top",
        fontsize=9.5, color=C_CBL, fontweight="bold")
ax.text(0.97, 0.20, r"ASCEND: $t \propto d_z^{-0.45}$",
        transform=ax.transAxes, ha="right", va="top",
        fontsize=9.5, color=C_ASCEND, fontweight="bold")

# (c) Speedup factor bar chart
ax = axes[0][2]
def speedup_along(df, axis):
    # Treat timeouts as their observed (censored) time — this is a lower
    # bound on the true CBL cost, so the resulting speedup is conservative.
    g = df.groupby([axis, "method"])["time_sec"].mean().unstack("method")
    g = g.reset_index().sort_values(axis)
    g["speedup"] = g["cbl"] / g["ascend"]
    return g

sp_dx = speedup_along(slice_dx, "d_x")
sp_dz = speedup_along(slice_dz, "d_z")

# Track which cells contained CBL timeouts so we can hatch those bars
to_dx = set(d[(d.timed_out) & (d.n == 1024) & (d.d_z == 10)].d_x.unique())
to_dz = set(d[(d.timed_out) & (d.n == 1024) & (d.d_x == 5)].d_z.unique())

bars, labels, colours, hatches = [], [], [], []
for _, r in sp_dx.iterrows():
    bars.append(r["speedup"])
    labels.append(rf"$d_x$={int(r['d_x'])}")
    colours.append("#7FB3D5")
    hatches.append("///" if int(r["d_x"]) in to_dx else "")
for _, r in sp_dz.iterrows():
    bars.append(r["speedup"])
    labels.append(rf"$d_z$={int(r['d_z'])}")
    colours.append("#1f6f9c")
    hatches.append("///" if int(r["d_z"]) in to_dz else "")

x = np.arange(len(bars))
for xi, h, col, hat in zip(x, bars, colours, hatches):
    ax.bar(xi, h, color=col, edgecolor="white" if not hat else C_TIMEOUT,
           linewidth=0.8 if not hat else 1.2, hatch=hat)
for xi, h in zip(x, bars):
    ax.text(xi, h * 1.07, f"{h:,.0f}×".replace(",", ","),
            ha="center", va="bottom", fontsize=9.0)
ax.set_xticks(x)
ax.set_xticklabels(labels, rotation=45, ha="right", fontsize=9.0)
ax.set_yscale("log")
ax.set_ylabel("Speedup  (CBL time / ASCEND time)")
ax.set_title("(c) ASCEND speedup factor")
ax.axhline(1, color=GREY, lw=0.7, linestyle="--")
ax.set_ylim(0.7, max(bars) * 4)
ax.grid(axis="y", which="major", alpha=0.25, lw=0.5)
ax.legend(handles=[
    Patch(facecolor="#7FB3D5", label=r"varying $d_x$"),
    Patch(facecolor="#1f6f9c", label=r"varying $d_z$"),
    Patch(facecolor="white", edgecolor=C_TIMEOUT, hatch="///",
          label="≥1 CBL timeout (lower bound)"),
], loc="upper left", fontsize=8.6)

# =========================== ROW 2 : QUALITY ==============================
# (d) Computational work — CI tests performed
ax = axes[1][0]
sweep_panel(ax, slice_dz, "d_z", "n_ci_tests", log_y=True)
ax.set_xlabel(r"$d_z$")
ax.set_ylabel("Conditional independence\ntests performed")
ax.set_title("(d) Computational work")
ax.set_xticks([10, 20, 30, 40, 50])
ax.grid(axis="y", which="major", alpha=0.25, lw=0.5)
ax.text(0.02, 0.97, r"n=1024,  $d_x$=5",
        transform=ax.transAxes, va="top", ha="left",
        fontsize=8.8, color=GREY)

# (e) Decisiveness — coverage (1 - NA fraction)
ax = axes[1][1]
sweep_panel(ax, slice_dz, "d_z", "coverage")
ax.set_xlabel(r"$d_z$")
ax.set_ylabel("Coverage  (1 − NA fraction)")
ax.set_title("(e) Decisiveness")
ax.set_xticks([10, 20, 30, 40, 50])
ax.set_ylim(0.45, 1.03)
ax.grid(axis="y", which="major", alpha=0.25, lw=0.5)
ax.text(0.02, 0.05, r"n=1024,  $d_x$=5",
        transform=ax.transAxes, va="bottom", ha="left",
        fontsize=8.8, color=GREY)

# (f) Accuracy: F1 vs d_x
ax = axes[1][2]
sweep_panel(ax, slice_dx, "d_x", "f1")
ax.set_xlabel(r"$d_x$")
ax.set_ylabel("F1 score")
ax.set_title(r"(f) Accuracy vs $d_x$")
ax.set_xticks([5, 10, 15, 20])
ax.set_ylim(-0.02, 1.03)
ax.grid(axis="y", which="major", alpha=0.25, lw=0.5)
ax.text(0.02, 0.05, r"n=1024,  $d_z$=10",
        transform=ax.transAxes, va="bottom", ha="left",
        fontsize=8.8, color=GREY)

# ---- shared legend (top) -------------------------------------------------
handles = [
    mlines.Line2D([], [], color=C_ASCEND, marker="o", lw=2.2, ms=7,
                  markeredgecolor="white", markeredgewidth=1.0,
                  label="ASCEND (ours)"),
    mlines.Line2D([], [], color=C_CBL, marker="o", lw=2.2, ms=7,
                  markeredgecolor="white", markeredgewidth=1.0,
                  label="CBL (baseline)"),
    mlines.Line2D([], [], color=C_TIMEOUT, marker="x", lw=0, ms=9,
                  markeredgewidth=2.4, label="CBL exceeded 1 h timeout"),
]
fig.legend(handles=handles, loc="upper center",
           bbox_to_anchor=(0.5, 0.965), ncol=3, frameon=False,
           fontsize=10.5)

fig.suptitle("ASCEND vs CBL  —  scaling, computational cost, and discovery quality "
             "(5 seeds, ribbons = ±1 SEM)",
             y=1.00, fontsize=12.5, fontweight="bold")

for ext in ("pdf", "png"):
    out = f"{OUT_DIR}/ascend_vs_cbl.{ext}"
    plt.savefig(out)
    print("wrote", out)